boskos/boskos.go

```diff
@@ -0,0 +1,304 @@
+		durationStr := req.URL.Query().Get("duration")
+		if job == "" || durationStr == "" {
+			logrus.Warning("[BadRequest]Job and duration must present in the request")
+			http.Error(res, "Job and duration must present in the request", http.StatusBadRequest)
+			return
+		}
+
+		duration, err := time.ParseDuration(durationStr)
+		if err != nil {
+			http.Error(res, "Invalid Duration", http.StatusBadRequest)
+			logrus.WithError(err).Infof("Bad duration: %v", durationStr)
+			return
+		}
+
+		logrus.Infof("Job %v request for a project for %v", job, durationStr)
+
+		proj, err := myRanch.GimmeCattle(job, duration)
+		if err != nil {
+			logrus.WithError(err).Info("Error from GimmeCattle")
+			http.Error(res, fmt.Sprintf("Failed request with error %v", err), http.StatusInternalServerError)
+			return
+		}
+
+		if proj != nil {
+			projJSON, err := json.Marshal(proj)
+			if err != nil {
+				logrus.WithError(err).Infof("json.Marshal failed : %v", proj)
+				http.Error(res, error.Error(err), http.StatusInternalServerError)
+				return
+			}
+			logrus.Infof("Job %v checked out %v", job, string(projJSON))
+			fmt.Fprint(res, string(projJSON))
+		} else {
+			logrus.Infof("No available project")
+			res.WriteHeader(http.StatusNoContent)
+			fmt.Fprint(res, "No available project, please try again later\n")
+		}
+	}
+}
+
+func handleReturnProject(myRanch *RanchoSantaFe) http.HandlerFunc {
+	return func(res http.ResponseWriter, req *http.Request) {
+		logrus.WithField("handler", "handleReturnProject").Infof("From %v", req.RemoteAddr)
+
+		// validate vars
+		job := req.URL.Query().Get("job")
+		if job == "" {
+			http.Error(res, "job must present in the request", http.StatusBadRequest)
+			return
+		}
+
+		logrus.Infof("Job %v returns a project", job)
+		if err := myRanch.ReturnCattle(job); err == nil {
+			fmt.Fprint(res, "Successful\n")
+		} else {
+			logrus.WithError(err).Info("Return failed")
+			http.Error(res, error.Error(err), http.StatusInternalServerError)
+		}
+	}
+}
+
+func handleHeartbeat(myRanch *RanchoSantaFe) http.HandlerFunc {
+	return func(res http.ResponseWriter, req *http.Request) {
+		logrus.WithField("handler", "handleHeartbeat").Infof("From %v", req.RemoteAddr)
+		if !myRanch.CheckLease() {
+			res.WriteHeader(http.StatusNoContent)
+		}
+	}
+}
+
+type RanchoSantaFe struct {
+	MyCattles    []*GCPProject
+	JobToCattles map[string]*GCPProject
+	CattleLock   sync.RWMutex
+	Storage      string
+}
+
+func NewProject(projName string) *GCPProject {
+	newProject := &GCPProject{
+		Name: projName,
+	}
+	return newProject
+}
+
+func NewRancho(config string, storage string) *RanchoSantaFe {
+	file, err := os.Open(config)
+	if err != nil {
+		logrus.WithError(err).Fatal(fmt.Sprintf("Fail to open config file %v.", config))
+	}
+	defer file.Close()
+
+	newRanch := &RanchoSantaFe{
+		Storage:      storage,
+		JobToCattles: make(map[string]*GCPProject),
+	}
+
+	if storage != "" {
+		buf, err := ioutil.ReadFile(storage)
```

Comment:
You can use a `json.Decoder` to avoid reading into a buffer, but that's not super important.